modules101 article (#57)
valkey-io/valkey-io.github.io: _posts/2024-04-26-modules-101.markdown

---
layout: post
title:  "Valkey modules 101"
authors: 
  - dmitrypol
date: 2024-04-26 01:01:01 -0700
categories: modules
---

## What are Valkey modules?  

> The idea of modules is to keep the core Valkey functionality lean and add extra features (such as new commands and data types) where needed.  Modules can be written in C or other languages that have C bindings.  In this article we will go through the process of building simple modules in C and Rust (using the Valkey Module Rust SDK).  This article expects the audience to be at least somewhat familiar with C, Rust and Valkey.  

Reviewer: ```suggestion
The idea of modules is to allow adding extra features (such as new commands and data types) to Valkey without making changes to the core code.  Modules are a special type of code distribution called a shared library, which can be loaded by other programs at runtime and executed.  Modules can be written in C or other languages that have C bindings.  In this article we will go through the process of building simple modules in C and Rust (using the Valkey Module Rust SDK).  This article expects the audience to be at least somewhat familiar with C, Rust and Valkey.  
```

@dmitrypol @stockholmux I would also like to suggest we enforce one sentence per line, it becomes much easier to review. 

The blog also assumes knowledge of git and cloning. I would suggest either adding more detailed structure for cloning Valkey so that it keeps the bar low.

Author: updated

Author: not sure what you mean by one sentence per line.  That will impact how the text is presented in the published blog.

Reviewer 2: @dmitrypol Create a new line for each sentence (not in a code block). Markdown create paragraph breaks after two new lines, so it'll look fine in the resulting blog post.

Example:
```
This is my first sentence. 
This is my second sentence.

This is a new paragraph.
```

Will render like:
> This is my first sentence. This is my second sentence.
> 
> This is a new paragraph.

@madolson I _really_ need to setup a style guide and perhaps a action to check for these. I'll add it to the ever growing to-do list.

Author: I thought it does it too but when I make the change locally and run `bundle exec jekyll serve` it generates the break after one line. I will make the change and hopefully it will look fine

> The idea of modules is to keep the core Valkey functionality lean and add extra features (such as new commands and data types) where needed.  Modules can be written in C or other languages that have C bindings.  In this article we will go through the process of building simple modules in C and Rust (using the Valkey Module Rust SDK).  This articles expects the audience to be at least somewhat familiar with C, Rust and Valkey.  

Reviewer 2: > This articles expects the audience to be

article (singular)

Author: udpated

## Hello World module in C

> If we clone [Valkey](https://github.com/valkey-io/valkey) repo we will find numerous examples in `src/modules`.  Let's create a new file `module1.c` in the same folder.  

Reviewer: ```suggestion
If we clone the [Valkey](https://github.com/valkey-io/valkey) repo we will find numerous examples in `src/modules`.  Let's create a new file `module1.c` in the same folder.  
```

Author: done

```c
#include "../valkeymodule.h"

int hello(ValkeyModuleCtx *ctx, ValkeyModuleString **argv, int argc) {
    VALKEYMODULE_NOT_USED(argv);
    VALKEYMODULE_NOT_USED(argc);
    return ValkeyModule_ReplyWithSimpleString(ctx, "world1");
}

int ValkeyModule_OnLoad(ValkeyModuleCtx *ctx, ValkeyModuleString **argv, int argc) {
    VALKEYMODULE_NOT_USED(argv);
    VALKEYMODULE_NOT_USED(argc);
    if (ValkeyModule_Init(ctx,"module1",1,VALKEYMODULE_APIVER_1) 
        == VALKEYMODULE_ERR) return VALKEYMODULE_ERR;
    if (ValkeyModule_CreateCommand(ctx,"module1.hello", hello,"",0,0,0) 
        == VALKEYMODULE_ERR) return VALKEYMODULE_ERR;
    return VALKEYMODULE_OK;
}
```

> All we are doing is initalizing `module1` and creating a command `hello` that returns `world1`.  Now we need to update `src/modules/Makefile`

Reviewer 2: initalizing -> initializing

Author: udpated

> All we are doing is initializing `module1` and creating a command `hello` that returns `world1`.  Now we need to update `src/modules/Makefile`

Reviewer: I suppose there is a question of what is 101 vs what isn't. "All we are doing is initializing module1" is doing a lot of heavy lifting in this sentence. What is the initialization code? What is creating the command? Do we want to explain that here or maybe in a follow up.

Author: added more details.

```makefile
all: ... module1.so

module1.xo: ../valkeymodule.h

module1.so: module1.xo
	$(LD) -o $@ $^ $(SHOBJ_LDFLAGS) $(LIBS) -lc
```

> Run `make` inside `src/modules` folder.  We will see numerous *.so and *.xo files created in `src/modules` including `module1.so`.  

Reviewer 2: put the patterns (*.so/*.xo) in backticks

Author: done

> Run `make` inside `src/modules` folder.  We will see numerous `*.so` and `*.xo` files created in `src/modules` including `module1.so`.  

Reviewer: ```suggestion
Run `make module1.so` inside `src/modules` folder.  This will compile our module in the src/modules folder. 
```

No need to generate everything.

Author: fair point, updated

## Hello World module in Rust

> We will create a new Rust package by running `cargo new --lib module2` in bash.  Inside the `module2` folder we will have `Cargo.toml` and `src/lib.rs` files.  To install valkey-module SDK run `cargo add valkey-module`.  It will create or update `Cargo.lock` file.  

Reviewer 2: To install **the** valkey-module SDK run cargo add valkey-module. It will create or update **the** Cargo.lock file.

Author: updated

> We will create a new Rust package by running `cargo new --lib module2` in bash.  Inside the `module2` folder we will have `Cargo.toml` and `src/lib.rs` files.  To install the valkey-module SDK run `cargo add valkey-module`.  It will create or update the `Cargo.lock` file.  

Reviewer: `cargo add valkey-module` is not a comand.

Reviewer: Yeah, I updated cargo and it worked. I apparently installed cargo in 2020

> Modify `Cargo.toml` to specify crate-type.  Read [Rust docs](https://doc.rust-lang.org/reference/linkage.html) to understand more about `crate-type`.

Reviewer: ```suggestion
Modify `Cargo.toml` to specify the crate-type to be "cdylib", which will tell cargo to build the target as a shared library .  Read [Rust docs](https://doc.rust-lang.org/reference/linkage.html) to understand more about `crate-type`.
```

Author: updated

```
[lib]
crate-type = ["cdylib"]
```

Now in `src/lib.rs` replace the existing code with the following:

```rust
#[macro_use]
extern crate valkey_module;

> use valkey_module::{Context, ValkeyResult, ValkeyString, ValkeyValue};

Reviewer: We are missing the Valkey dependency

> use valkey_module::{Context, ValkeyResult, ValkeyString, ValkeyValue};

Reviewer: Right now I get an issue building this, 

```
error[E0432]: unresolved import `valkey_module::ValkeyValue`
 --> src/lib.rs:4:58
  |
4 | use valkey_module::{Context, ValkeyResult, ValkeyString, ValkeyValue};
  |                                                          ^^^^^^^^^^^ no `ValkeyValue` in the root

For more information about this error, try `rustc --explain E0432`.
```

Seems like Valkeyvalue is still called a RedisModule in the published version.

Author: This needs valkey-module 0.1.0 crate which I will publish to crates.io once you approve https://github.com/valkey-io/valkeymodule-rs/pull/8.  I can publish it now as is but then if we make changes to the SDK we will need to bump version.

fn hello(_ctx: &Context, _args: Vec<ValkeyString>) -> ValkeyResult {
    Ok(ValkeyValue::SimpleStringStatic("world2"))
}

valkey_module! {
    name: "module2",
    version: 1,
    allocator: (valkey_module::alloc::ValkeyAlloc, valkey_module::alloc::ValkeyAlloc),
    data_types: [],
    commands: [
        ["module2.hello", hello, "", 0, 0, 0],
    ]
}
```

> Rust syntax is a bit different than C but we are creating `module2` with command `hello` that returns `world2` string.  We are using external crate `valkey_module` with [Rust macros](https://doc.rust-lang.org/book/ch19-06-macros.html) and passing it variables like `name` and `version`.  Some variables like `data_types` and `commands` are arrays and we can pass zero, one or many values.  Since we not using ctx or args we prefix them with `_` (Rust convention) instead of using `VALKEYMODULE_NOT_USED` as we did in C.  

Reviewer 2: "Since we not using ctx or args..." -> "Since we're not using ctx or args..."

Author: updated

Run `cargo build` in the root folder.  We will now see `target/debug/libmodule2.dylib` (on macOS).  Build will produce *.so files on Linux and *.dll files on Windows.  

## Run Valkey server with both modules

> Go back into Valkey repo folder and run make to compile the Valkey code.  Then add these lines to the bottom of `valkey.conf` file.

Reviewer 2: Go back into **the** Valkey repo folder and run make to compile the Valkey code.  Then add these lines to the bottom of  **the** `valkey.conf` file.

Author: updated

> Go back into the Valkey repo folder and run make to compile the Valkey code.  Then add these lines to the bottom of the `valkey.conf` file.

Reviewer: ```suggestion
Go back into the Valkey repo folder and run `make` to compile the Valkey code.  Then add these lines to the bottom of the `valkey.conf` file.
```

Author: updated

```
loadmodule path_to_valkey/src/modules/module1.so

> loadmodule path_to_module2/target/debug/libmodule2.dylib

Reviewer: Should probably clarify the "path_to_*" needs to be updated.

Author: updated

```

and run `src/valkey-server valkey.conf`.  You will see these messages in the log output.  

```
Module 'module1' loaded from path_to_valkey/src/modules/module1.so
...
Module 'module2' loaded from path_to_module2/target/debug/libmodule2.dylib
```

Then use `src/valkey-cli` to connect.  

```bash
src/valkey-cli -3
127.0.0.1:6379> module list
1) 1# "name" => "module2"
   2# "ver" => (integer) 1
   3# "path" => "path_to_module2/target/debug/libmodule2.dylib"
   4# "args" => (empty array)
2) 1# "name" => "module1"
   2# "ver" => (integer) 1
   3# "path" => "path_to_valkey/src/modules/module1.so"
   4# "args" => (empty array)
127.0.0.1:6379> module1.hello
world1
127.0.0.1:6379> module2.hello
world2
```

> We can now run both modules side by side and if we modify either C or RS file, recompile the code and restart `valkey-server` we will get the new functionality.  Alternatively to specifying modules in .conf file we can use MODULE LOAD and UNLOAD from valkey-cli at runtime.  

Reviewer: ```suggestion
We can now run both modules side by side and if we modify either C or RS file, recompile the code and restart `valkey-server` we will get the new functionality.  Alternatively to specifying modules in .conf file we can use MODULE LOAD and UNLOAD from valkey-cli at runtime.  
```

I don't like blogs that say, "Alternatively do X/Y" without explaining it. So would prefer to either drop the reference or show an example of it.

---
layout: post
title:  "Valkey modules 101"
authors: 
  - dmitrypol
date: 2024-04-26 01:01:01 -0700

> categories: update

Reviewer: What are our categories?

Author: good question.  I updated the category to `modules` but I can omit it.  It becomes part of the URL so we don't want to change it afterwards or it will direct links

---

> ## What are Valkey (or Redis) modules?  

Reviewer: I think for branding purposes, we should omit Redis completely from this post, unless specifically talking about the license changes and why the project exists.

Author: Completely agree on branding.  I removed all references except the ones required by SDK code.  Working to update SDK but that's going to take a bit of time.

Author: updated blog to use Valkey* vars.  You can see SDK changes in https://github.com/valkey-io/valkeymodule-rs/pull/8

```c
#include "../valkeymodule.h"

int hello(ValkeyModuleCtx *ctx, ValkeyModuleString **argv, int argc) {
    VALKEYMODULE_NOT_USED(argv);
    VALKEYMODULE_NOT_USED(argc);
    return ValkeyModule_ReplyWithSimpleString(ctx, "world1");
}

int ValkeyModule_OnLoad(ValkeyModuleCtx *ctx, ValkeyModuleString **argv, int argc) {
    VALKEYMODULE_NOT_USED(argv);
    VALKEYMODULE_NOT_USED(argc);
    if (ValkeyModule_Init(ctx,"module1",1,VALKEYMODULE_APIVER_1) 
        == VALKEYMODULE_ERR) return VALKEYMODULE_ERR;
    if (ValkeyModule_CreateCommand(ctx,"module1.hello", hello,"",0,0,0) 
        == VALKEYMODULE_ERR) return VALKEYMODULE_ERR;
    return VALKEYMODULE_OK;
}
```

All we are doing is initalizing `module1` and creating a command `hello` that returns `world1`.  Now we need to update `src/modules/Makefile`

```makefile
all: ... module1.so

module1.xo: ../valkeymodule.h

module1.so: module1.xo
	$(LD) -o $@ $^ $(SHOBJ_LDFLAGS) $(LIBS) -lc
```

Run `make` inside `src/modules` folder.  We will see numerous *.so and *.xo files created in `src/modules` including `module1.so`.  

## Hello World module in Rust

> If we clone the [valkeymodule-rs](https://github.com/valkey-io/valkeymodule-rs) repo we will find `examples` folder with numerous .rs files.  Valkeymodule-rs is a recent fork of redismiodule-rs and there are still references to Redis in it.  We plan to update this repo and add it to [crates.io](https://crates.io/) but for now we will use it as is.  

Reviewer: Ideally we would use the valkeymodule-rs SDK correctly in this blog post, which is to take a dependency on it and then build it. We don't really want people using the sdk this way, and in the long term people might get confused by having this legacy post.

Reviewer: Can we also cleanup the references to redis before publishing this? I think it would be better for branding to limit the references to Redis in this post.

Author: Agree on branding.  I am working to get SDK into crates.io but I want to cleanup before publishing it as crate.

Author: @madolson - per our chat I published https://crates.io/crates/valkey-module and updated this blog.

Let's create `examples/module2.rs`

```rust
use redis_module::{redis_module, Context, RedisResult, RedisString, RedisValue};

fn hello(_ctx: &Context, _args: Vec<RedisString>) -> RedisResult {
    Ok(RedisValue::SimpleStringStatic("world2"))
}

redis_module! {
    name: "module2",
    version: 1,
    allocator: (redis_module::alloc::RedisAlloc, redis_module::alloc::RedisAlloc),
    data_types: [],
    commands: [
        ["module2.hello", hello, "", 0, 0, 0],
    ],
}
```

> Rust syntax is a bit different than C but we are creating `module2` with command `hello` that returns `world2`.  Now we need to update `Cargo.toml` in the root folder.  Add this the bottom of the file:

Reviewer: It would probably be useful to explain what this is doing, not just say that "the syntax is different".

Author: added details

To install the valkey-module SDK run `cargo add valkey-module` inside `module2` folder.  
Alternativley we can add `valkey-module = "0.1.0` in `Cargo.toml` under `[dependencies]`.  
Run `cargo build` and it will create or update the `Cargo.lock` file.  

Modify `Cargo.toml` to specify the crate-type to be "cdylib", which will tell cargo to build the target as a shared library.  
Read [Rust docs](https://doc.rust-lang.org/reference/linkage.html) to understand more about `crate-type`.

```
[lib]
crate-type = ["cdylib"]
```

Now in `src/lib.rs` replace the existing code with the following:

```rust
#[macro_use]
extern crate valkey_module;

use valkey_module::{Context, ValkeyResult, ValkeyString, ValkeyValue};

fn hello(_ctx: &Context, _args: Vec<ValkeyString>) -> ValkeyResult {
    Ok(ValkeyValue::SimpleStringStatic("world2"))
}

valkey_module! {
    name: "module2",
    version: 1,
    allocator: (valkey_module::alloc::ValkeyAlloc, valkey_module::alloc::ValkeyAlloc),
    data_types: [],
    commands: [
        ["module2.hello", hello, "", 0, 0, 0],
    ]
}
```

Rust syntax is a bit different than C but we are creating `module2` with command `hello` that returns `world2` string.  

> We are using external crate `valkey_module` with [Rust macros](https://doc.rust-lang.org/book/ch19-06-macros.html) and passing it variables like `name` and `version`.  

Reviewer: ```suggestion
We are using the external crate `valkey_module` with [Rust macros](https://doc.rust-lang.org/book/ch19-06-macros.html) and passing it variables like `name` and `version`.  
```

Author: updated

Some variables like `data_types` and `commands` are arrays and we can pass zero, one or many values.  
Since we are not using ctx or args we prefix them with `_` (Rust convention) instead of `VALKEYMODULE_NOT_USED` as we did in C.  

Run `cargo build` in the root folder.  
We will now see `target/debug/libmodule2.dylib` (on macOS).  

> Build will produce *.so files on Linux and *.dll files on Windows.  

Reviewer: ```suggestion
The build will produce *.so files on Linux and *.dll files on Windows.  
```

Author: updated

Go back into the Valkey repo folder and run `make` to compile the Valkey code.  
Then add these lines to the bottom of the `valkey.conf` file.

```
loadmodule UPDATE_PATH_TO_VALKEY/src/modules/module1.so
loadmodule UPDATE_PATH_TO_MODULE2/target/debug/libmodule2.dylib
```

and run `src/valkey-server valkey.conf`.  
You will see these messages in the log output.  

```
Module 'module1' loaded from UPDATE_PATH_TO_VALKEY/src/modules/module1.so
...
Module 'module2' loaded from UPDATE_PATH_TO_MODULE2/target/debug/libmodule2.dylib
```

Then use `src/valkey-cli` to connect.  

```bash
src/valkey-cli -3
127.0.0.1:6379> module list
1) 1# "name" => "module2"
   2# "ver" => (integer) 1
   3# "path" => "UPDATE_PATH_TO_MODULE2/target/debug/libmodule2.dylib"
   4# "args" => (empty array)
2) 1# "name" => "module1"
   2# "ver" => (integer) 1
   3# "path" => "UPDATE_PATH_TO_VALKEY/src/modules/module1.so"
   4# "args" => (empty array)
127.0.0.1:6379> module1.hello
world1
127.0.0.1:6379> module2.hello
world2
```

We can now run both modules side by side and if we modify either C or RS file, recompile the code and restart `valkey-server` we will get the new functionality.  

> Alternatively to specifying modules in `valkey.conf` file we can use MODULE LOAD and UNLOAD from `valkey-cli`.  

Reviewer: ```suggestion
As an alternative to specifying modules in `valkey.conf` file, we can use `MODULE LOAD` and `UNLOAD` from `valkey-cli` to update the server.
```

Author: updated
